service_container/service_subscribers.rst

```diff
@@ -0,0 +1,403 @@
+.. index::
+single: DependencyInjection; Service Subscribers
+
+Service Subscribers
+================
+
+Symfony's :doc:`Service Locators </service_container/service_locators>` provide
+a powerful way of passing a subset of services from the service container to a
+service. Sometimes, you may want a more concrete way of defining the services
+contained within a service locator. By implementing
+:class:`Symfony\\Component\\DependencyInjection\\ServiceSubscriberInterface`
+you can specify the contents of the service locator from the object itself.
+This is useful when a set of services have the same parent class and similar
+dependencies.
+
+Suppose you've got a mailing system with multiple ``Updater`` classes for
+sending out mails on different occasions. Since every updater has the same
+purpose, we start with a base updater::
+
+    // src/AppBundle/Updates/AbstractUpdater.php
+    namespace AppBundle\Updates;
+
+    use Twig\Environment;
+
+    abstract class AbstractUpdater
+    {
+        protected $mailer;
+        protected $twig;
+
+        public function __construct(\Swift_Mailer $mailer, Environment $twig)
+        {
+            $this->mailer = $mailer;
+            $this->twig = $twig;
+        }
+
+        abstract public function update($entity);
+
+        /**
+         * Renders a view with Twig
+         */
+        public function render($template, $parameters)
+        {
+            return $this->twig->render($template, $parameters);
+        }
+
+        /**
+         * Sends an email with Swiftmailer
+         */
```

Comment:
This example contains some docblocks, I'm not sure how useful they are for the actual example as the code underneath is explains pretty much the same thing

Reply:
True, they're from a time where the methods weren't just simple one-liners. Thanks.